ios/MaasCore/Sources/MaasCore/Extensions/Publishers+Retry.swift

```diff
@@ -0,0 +1,106 @@
+import Combine
+
+// inspiration: https://www.wwt.com/article/creating-your-own-custom-combine-operator
+
+extension Publishers {
+
+    // A publisher that attempts to recreate its subscription to a failed upstream publisher.
+    public struct RetryWithRecovery<Upstream: Publisher, RecoveryPublisher: Publisher>: Publisher {
+        
+        public typealias Output = Upstream.Output
+        
+        public typealias Failure = Upstream.Failure
+        
+        let upstream: Upstream
+        let failureShouldRetry: (Failure) -> Bool
+        let recoveryPublisher: RecoveryPublisher
+        
+        public init(upstream: Upstream, failureShouldRetry: @escaping (Failure) -> Bool, recoveryPublisher: RecoveryPublisher) {
+            self.upstream = upstream
+            self.failureShouldRetry = failureShouldRetry
+            self.recoveryPublisher = recoveryPublisher
+        }
+        
+        public func receive<S: Subscriber>(subscriber: S) where Upstream.Failure == S.Failure, Upstream.Output == S.Input {
+            self.upstream
+                .catch { e -> AnyPublisher<Output, Failure> in
+                    guard failureShouldRetry(e) else {
+                        return Fail<Output, Failure>(error: e)
```

Comment:
Cant we use `Just(e)` ? 🤔

Reply:
`Just` creates a publisher with output, we need a publisher with failure -> `Fail`
we can simplify this line to `return Fail(error: e)`